Add extra check for a request header key to EndTransaction and HeartbeatTxn

kkaneda opened the pull request.

kkaneda committed:
Add extra check for commands that update transaction keys

The check is added in ExecuteCmd() to prevent bad requests from coming
in an early stage. Add a test for bad key requests.

vivekmenezes commented on storage/replica_command.go:
```diff
@@ -250,6 +250,10 @@ func (r *Replica) EndTransaction(batch engine.Engine, ms *engine.MVCCStats, args
 	if args.Txn == nil {
 		return reply, nil, util.Errorf("no transaction specified to EndTransaction")
 	}
+	if !bytes.Equal(args.Key, args.Txn.Key) {
```
why do we care?

kkaneda replied:
I might not be understanding code correctly, but I think we need this to avoid processing commands (e.g., `HeartbeatTxn` and `EndTransaction`) for the same txn at the same time. The command queue looks at `args.Key` to block a command whose keys overlap with already-running commands.

vivekmenezes replied:
Is it possible for us to check this immediately on receiving the request. That way you need to check it only once.

kkaneda replied:
I see. I think that's doable. Let me take a look.

vivekmenezes commented on storage/replica_command.go:
```diff
@@ -646,6 +650,13 @@ func (r *Replica) RangeLookup(batch engine.Engine, args proto.RangeLookupRequest
 func (r *Replica) HeartbeatTxn(batch engine.Engine, ms *engine.MVCCStats, args proto.HeartbeatTxnRequest) (proto.HeartbeatTxnResponse, error) {
 	var reply proto.HeartbeatTxnResponse
 
+	if args.Txn == nil {
```
this is coming from another node? Are you protecting against a bad sending node?

kkaneda replied:
Yes, that was my understanding. I thought the check is nice to have to avoid potential bugs.

vivekmenezes replied:
Perhaps we should write some tests that basically recreate all kinds of bad sender scenarios, and then add code that protects us. That could be a slightly more ambitious change. Thanks for taking this on!

kkaneda replied:
Sure, will take a look this as well. Thanks for the review!

vivekmenezes commented on storage/replica_command.go:
```diff
@@ -646,6 +650,13 @@ func (r *Replica) RangeLookup(batch engine.Engine, args proto.RangeLookupRequest
+	if args.Txn == nil {
+		return reply, util.Errorf("no transaction specified to HeartbeatTxn")
+	}
+	if !bytes.Equal(args.Key, args.Txn.Key) {
```
do we care?

bdarnell commented:
LGTM

kkaneda commented:
I've updated the PR. Could you take another look, @vivekmenezes ? I've moved a check to `Store.executeCmd`. 

I've defined `isTxnKeyUpdate` in `api.go`, but it might be overkill since it is used only in one place.

vivekmenezes commented on storage/store.go:
```diff
@@ -144,6 +144,39 @@ func verifyKeys(start, end proto.Key, checkEndKey bool) error {
+	case *proto.EndTransactionRequest:
+		txn = tArgs.Txn
+	case *proto.HeartbeatTxnRequest:
+		txn = tArgs.Txn
```
Combine these two into:

case *proto.EndTransactionRequest, *proto.HeartbeatTxnRequest:

tschottdorf replied:
that's not possible since that leaves `tArgs` with two types.

vivekmenezes replied:
oh blimey

tamird commented on storage/store.go:
```diff
@@ -144,6 +144,39 @@ func verifyKeys(start, end proto.Key, checkEndKey bool) error {
+		return nil
+	}
+	var txn *proto.Transaction
+	switch tArgs := args.(type) {
```
should `{ResolveIntent,ResolveIntentRange}` be covered here?
with these additions, can the checking in `(*Replica).{EndTransaction,ResolveIntent,ResolveIntentRange}` be removed?

kkaneda replied:
Hmm, we can, but we need to check only `txn != nil` as keys in `RequestHeader` are transaction keys. I could be convinced either way, but I thought that adding too many checks in `executeCmd` might be overkill.

Yes, we can remove the checking in `EndTransaction`, but I wasn't sure either. It might be still nice to have the checks inside each command.

tamird replied:
> It might be still nice to have the checks inside each command.

Does it produce different behaviour?

tschottdorf replied:
just a heads up that `ResolveIntent` will no longer use the `Header.Txn` field in the near future, so it'll always have to be checked in the actual command.

kkaneda replied:
No, the behavior will be the same. `Store.executeCmd

tamird replied:
Then there's no point duplicating the check.

kkaneda replied:
Moved the check to each command..

vivekmenezes commented on storage/store_test.go:
```diff
@@ -1429,6 +1429,70 @@ func TestRaftNodeID(t *testing.T) {
+		} else {
+			log.Infof("err=%s", err)
+		}
+	}
```
Can this combined with the the iteration above?

kkaneda replied:
Sure, done!

vivekmenezes commented on storage/store.go:
```diff
@@ -144,6 +144,39 @@ func verifyKeys(start, end proto.Key, checkEndKey bool) error {
+// queue ensures that.)
+func verifyTransactionKeys(args proto.Request) error {
+	header := args.Header()
+	if !proto.IsTransactionKeyUpdate(args) {
```
Do you really need this? You can return in the default clause below right?

kkaneda replied:
Right, this was overkill.. Reverted the change to api.proto.

tschottdorf commented on storage/store_test.go:
```diff
@@ -1429,6 +1429,70 @@ func TestRaftNodeID(t *testing.T) {
+	testCases = append([]proto.Request{}, &tArgs0, &tArgs1, &tArgs2, &tArgs3, &tArgs4)
+	for i, test := range testCases {
+		if _, err := store.ExecuteCmd(context.Background(), test); err == nil {
+			t.Errorf("%d unexpected success of request %s", i, test)
```
could you match the actual errors here?

tschottdorf commented:
> I've defined isTxnKeyUpdate in api.go, but it might be overkill since it is used only in one place.

I agree. Why move a handful of low-level checks so high up the stack that you really need to know where to look? I'd prefer having `EndTransaction` and its companions self-contained. If anything, they could share an error for missing Txns so that your test cases simplify. At the end of the day, they're all just assertions anyways, not expected in practice.
The test cases look useful though.

kkaneda commented:
Sure, moved the check back to each command and reverted the change to api.go. Sorry for the back-and-forth.. PTAL.

tschottdorf commented:
LGTM. Thanks Kenji!

vivekmenezes commented:
LGTM. Sorry about the back and forth on this PR.
